sql: unexpected type *tree.DVoid for AsJSON

roachtest.sqlsmith/setup=rand-tables/setting=no-mutations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5645962?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5645962?buildTab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations) on master @ [7a8cac9b11036dcf401453eb36e531fe381b70b4](https://github.com/cockroachdb/cockroach/commits/7a8cac9b11036dcf401453eb36e531fe381b70b4):


```
test artifacts and logs in: /artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:265,sqlsmith.go:325,test_runner.go:896: error: pq: internal error: row_to_json(): unexpected type *tree.DVoid for AsJSON
		stmt:
		WITH
			with_1974 (col_12293) AS (SELECT * FROM (VALUES (NULL)) AS tab_4913 (col_12293)),
			with_1975 (col_12294)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								((-0.5289916396141052):::FLOAT8),
								(0.12303011864423752:::FLOAT8),
								(0.23907285928726196:::FLOAT8),
								((-0.3916063606739044):::FLOAT8),
								(NULL),
								((-0.7851547598838806):::FLOAT8)
						)
							AS tab_4914 (col_12294)
				)
		SELECT
			row_to_json(('':::VOID,))::JSONB AS col_12295
		FROM
			with_1974 AS cte_ref_583
		WHERE
			true
		LIMIT
			76:::INT8;
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-17267

--- a/pkg/sql/logictest/testdata/logic_test/void
+++ b/pkg/sql/logictest/testdata/logic_test/void
@@ -6,6 +6,13 @@ SELECT 'this will be ignored'::void
 ----
 ·
 
+# Regression test for #83791. row_to_json on a VOID should produce an
+# empty string, unadorned with single quotes, as in done in postgres.
+query T
+SELECT row_to_json((''::VOID, null))::JSONB AS col_12295
+----
+{"f1": "", "f2": null}
+
 query T
 select row (''::void, 2::int)
 ----

--- a/pkg/sql/sem/tree/datum.go
+++ b/pkg/sql/sem/tree/datum.go
@@ -3696,6 +3696,8 @@ func AsJSON(
 		return json.FromSpatialObject(t.Geometry.SpatialObject(), geo.DefaultGeoJSONDecimalDigits)
 	case *DGeography:
 		return json.FromSpatialObject(t.Geography.SpatialObject(), geo.DefaultGeoJSONDecimalDigits)
+	case *DVoid:
+		return json.FromString(AsStringWithFlags(t, fmtRawStrings)), nil
 	default:
 		if d == DNull {
 			return json.NullJSONValue, nil

